### ISSUE TITLE
:m0 moves line to second line instead of first

### ISSUE THREAD

**Issue by bschwehn** [jaredpar/VsVim#1671]
in vim 

:m0 will move the current line to the top. In vsvim it moves it to the second line, just below the top line. 

Is this a bug or do I have some setting wrong/another add in interfering? 

Thanks, 
Ben

**Comment by jaredpar** [comment]
That just looks like a bug in the `:m` implementation.  I will get that looked at.

**Comment by mattrrichard** [comment]
I dug into this, and I see what happening: When the interpreter executes a MoveTo or CopyTo command, it uses its `GetLine` method to convert to destination line specifier into an `ITextSnapshotLine`.  While in that method, it runs into one of the mismatches between Vim and VS's functionalities, how lines are numbered.  The `Util.VimLineToTssLine` method is used to convert and you lose the information that the user wanted to insert before the first line.

```
/// Vim is fairly odd in that it considers the top line of the file to be both line numbers
    /// 1 and 0.  The next line is 2.  The editor is a zero based index though so we need
    /// to take that into account
    let VimLineToTssLine line = 
        match line with
        | 0 -> 0
        | _ -> line - 1
```

The problem really boils down to losing information while calling GetLine: **a line specifier for line 1 and a line specifier for line 0 have the same result **.  

The only idea I have for how to fix this is to refactor `GetLineCore` into two methods where the first returns the target *Vim* line number and have CopyTo and MoveTo decide whether to insert at the start of the destination line or the end by looking at the target line number.

**Comment by ricksladkey** [comment]
My understanding was that in vim, AFTER line 0 and BEFORE line 1 refer to the same place.  So even though line numbers are one-based, operations that conceptually append can still insert before the first line due to the fictitious line zero.

**Comment by mattrrichard** [comment]
I don't think VisualStudio has a concept of the "fictitious line zero" which is the origin of the problem.

**Comment by jhamm** [comment]
I verified it works by building from the source. Can this issue be closed? This will help get us closer to 2.0.2 :smile:
